### ISSUE TITLE
Implement Image Classification on dermaMNIST

### ISSUE THREAD

**Issue by bdzyubak** [bdzyubak/torch-control#20]
As I have mostly done image segmentation in the past, let's do a classification project! 

dermaMNIST is a publicly available dataset of small RGB images of skin tumors with multi-label disease classifications. Unlike the hand digit dataset which is easy, the benchmark accuracy published in Nature with resnet50 is only 0.73. https://www.nature.com/articles/s41597-022-01721-8/tables/4

Let's see if we can do better. 
pip install medmnist
from medmnist import DermaMNIST
![image](https://github.com/bdzyubak/torch-control/assets/37943739/b1199493-d46a-4e03-b6f3-8f7be6a2bded)

**Comment by bdzyubak** [comment]
A basic tutorial style CNN + FCN gets to 0.69 train acc in one epoch and then stays there. 

ResNet50 from torchvision gets: 
Fresh weights: [98/100] train_loss: 0.333 - train_acc: 0.884 - eval_loss: 1.127 - eval_acc: 0.704 
Pretrained weights:

**Comment by bdzyubak** [comment]
For fun, I added more experiment options to: 
D:\Source\torch-control\projects\ComputerVision\dermMNIST\train_basic_network.py

A) I made a very wide model consisting of 4 CNN layers (channels=[256, 512, 1024, 2056]) and 2 FC blocks (channels = [10240, 10240]). The model is 2 GB on disk (compared to 98 MB for ResNet50)  and only achieves 0.69 accuracy! 

model = ConvNet(im_width=im_width, im_height=im_height, num_classes=num_classes, input_channels=input_channels,
                        cnn_start_channels=128, preclassifier_channels=10240)
                        
B) On the other hand, a deeper network with 4 * double CNN [256,...] + maxpool followed by 4 1024-channel FC layers get
 model = KaggleSimpleNet(im_width=im_width, im_height=im_height, num_classes=num_classes, input_channels=input_channels)
Fits the train data really well (though without dropout or augmentation, it doesn't generalize well). 
[100/100] train_loss: 0.039 - train_acc: 0.987 - eval_loss: 2.158 - eval_acc: 0.698

**Comment by bdzyubak** [comment]
The issue of the network underfitting this data was caused by a bug in the basic implementation. Pytorch CrossEntropyLoss applies nn.log_softmax() inside it and needs to be passed raw logits. If it is passed nn.log_softmax(), it works fine, but if it is passed nn.Softmax, this really hurts optimization
With nn.Softmax() activation - stuck at 0.67 train acc:
[22/100] train_loss: 1.499 - train_acc: 0.670 - eval_loss: 1.467 - eval_acc: 0.669

No nn.Softmax() layer - just pass logits
[22/100] train_loss: 0.129 - train_acc: 0.956 - eval_loss: 1.831 - eval_acc: 0.699

In the end, fitting the train data in dermaMNIST turned out to be very easy. There is still a class imbalance and generalizability issue to val data, which will be addressed in a future issue.
